packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps} from '@react-types/shared';
+import {getFocusableTreeWalker} from '@react-aria/focus';
+import {HTMLAttributes, MutableRefObject} from 'react';
+import {isElementVisible} from '@react-aria/focus/src/isElementVisible';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export interface AriaLandmarkProps  extends AriaLabelingProps {
+  role: 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary'
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+let landmarkRoles = ['main', 'region', 'search', 'navigation', 'form', 'banner', 'contentinfo', 'complementary'];
+const LANDMARK_ELEMENT_SELECTOR = landmarkRoles.map(value => `[role=${value}]`).join(',');
+
+function getLandmarkTreeWalker(root: HTMLElement, opts?) {
+  let selector = LANDMARK_ELEMENT_SELECTOR;
+  let walker = document.createTreeWalker(
+    root,
+    NodeFilter.SHOW_ELEMENT,
+    {
+      acceptNode(node) {
+        // Not sure we need this, can landmarks be nested?
+        // Skip nodes inside the starting node.
+        // if (opts?.from?.contains(node)) {
+        //   return NodeFilter.FILTER_REJECT;
+        // }
+
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && opts?.registeredLandmarks.includes(node)) {
+          return NodeFilter.FILTER_ACCEPT;
+        }
+
+        return NodeFilter.FILTER_SKIP;
+      }
+    }
+  );
+
+  if (opts?.from) {
+    walker.currentNode = opts.from;
+  }
+
+  return walker;
+}
+
+function treeDistance(child: HTMLElement, target: HTMLElement): number {
+  let distance = 0;
+  let node = child;
+  while (target && node.parentElement && node !== target) {
+    node = node.parentElement;
+    distance++;
+  }
+  return distance;
+}
+
+function hasLabel(ref: MutableRefObject<HTMLElement>) {
```

Comment:
Probably better if we can track it through the landmark registration?
We could add it to the landmarks Map

Reply:
So something like changing
`landmarks: Map<MutableRefObject<HTMLElement>, HTMLElement>`
to
`landmarks: Map<MutableRefObject<HTMLElement>,{label:string, lastFocused: HTMLElement}>` ?

Reply:
or based on [this comment](https://github.com/adobe/react-spectrum/pull/2750/files#r788239348), we probably actually want something like this:
```ts
landmarks: Array<{
    ref: MutableRefObject<HTMLElement>,
    role: string,
    label: string,
    lastfocused: HTMLElement
  }>
```